### PR TITLE
Attendance v1.0

### LINKED ISSUE #63: View Synced  Attendance Information

### PR DESCRIPTION
**Pull request by Scott123180**
- Record attendance 
- View attendance details
- View last 20 classes
- View Attendance Records for any day
- Edit participant information
- Search For Participants To Add
- Add new participant in system

Fixes #63

### CODE REVIEW

**Comment by Scott123180** [comment]
@vmjoseph @blackjack26 or @sami10015 any update on this? i know it's a lot of code to review and everyone is pretty busy around this time of the year. Just want to make sure it's integrated by tomorrow night.

**Comment by blackjack26** [comment]
@Scott123180. So @michellecrawley and I have been updating this branch and have actually pushed up changes to it (and rebased it onto the current version of develop). The things we changed were mostly styling and a few functional things. More specifically things like:

- Button colors
- Form look & feel
- Removing full-page displays (limited it to a max-width)
- Fixed the refresh issue when creating an attendance
- Fixed the attendance submission (was missing classid)
- Updated some code styles here and there

If you look through the code you can probably see the changes (also make sure to do a `git reset --hard origin/63-view-synced-attendance` to get these changes as they have been squashed into your original commit). If you wanna talk about the things we changed let us know. The majority of what you had was good and the flow of the pages has not been changed.

**Comment by Scott123180** [comment]
@blackjack26 Sounds good, thanks for the update :smile:

File: public/js/attendance-scripts/attendance-form-add-new-person.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    if(!validateZip(zip)){
+        return "Please enter a valid number zip";
+    }
+    //success
+    return true;
+}
+
+//first or last name
+function validateName(name) {
```
Comment (ghost) [inline public/js/attendance-scripts/attendance-form-add-new-person.js:89]: Valid first and last names are capable of having the `'` character. Ex: O'Reilly

File: public/js/attendance-scripts/attendance-form-add-new-person.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    if(isNaN(ageNumber)) return false;
+    if(typeof(ageNumber) !== "number") return false;
+    //returns true if age in valid range 18-100
+    return((ageNumber >= 18) && (ageNumber <= 100));
+}
+
+function validateNumChildren(num) {
+    var number = parseInt(num);
+    if(isNaN(number)) return false;
+    if(typeof(number) !== "number") return false;
```
Comment (ghost) [inline public/js/attendance-scripts/attendance-form-add-new-person.js]: I believe `parseInt` can only return a number or `NaN` which you've got covered

File: views/attendance/attendanceREADME.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#Attendance Documentation
```
Comment (ghost) [inline views/attendance/attendanceREADME.md]: Need to throw a space in front of `#` for markdown to render those headers (at least on GitHub specifically)

File: views/attendance/attendance_form.php [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+                "mi"            => $mi,
+                "ln"            => $ln,
+                "dob"           => $dob,
+                "zip"           => $zip,
+                "numChildren"   => $nc,
+                "race"          => $race,
+                "comments"      => null,
+                "present"       => true,
+                "isNew"         => $isNew, //isNew field from DB
+                //people who haven't completed the intake forms and just filled out info in the "no intake form" section
```
Comment (ghost) [inline views/attendance/attendance_form.php:139]: I don't think there is a `firstClass` . `isNew` is in ParticipantClassAttendance and could happen before (edge case) and after Intake packets as far as I know.

File: views/attendance/attendance_utilities.php [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+//input: zip code
+//output: true if valid 5 digit US zip
+function validateZip($zip) {
+    if(empty($zip)){
+        return false;
+    } else {
+        //validate zip code (from stackoverflow)
+        return preg_match("/(^\d{5}$)|(^\d{5}-\d{4}$)/", $zip);
```
Comment (ghost) [inline views/attendance/attendance_utilities.php]: Looks this returns true for extended Zip codes as well. I don't think these are ideal in `ZipCodes` or `ParticipantClassAttendance`

https://en.wikipedia.org/wiki/ZIP_Code#ZIP.2B4

File: views/attendance/edit_participant.php [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+            //returns true if a valid number of children
+            return((num >= 0) && (num <= 25))
+        }
+
+        function validateZip(zip) {
+            if (zip === "")
+                return true;
+
+            //validate zip code (from stackoverflow)
+            return (/(^\d{5}$)|(^\d{5}-\d{4}$)/.test(zip));
```
Comment (ghost) [inline views/attendance/edit_participant.php]: Same Zip+4 issue here

File: views/attendance/attendance_utilities.php [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+function validateCurriculum($results, $input){
+    while($row = pg_fetch_assoc($results)){
+        if($input == $row['topicname']) return true;
+    }
+    return false;
+}
+
+function validateDate($date){
+
```
Comment (ghost) [inline views/attendance/attendance_utilities.php]: Couple empty functions here. You might want a comment and `throw` an error "Function not defined" or just remove them for the time being.

File: views/attendance/historical_class_view.php [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+else if(isset($_POST["whichButtonHistoricalSearch"])){ //form historical search tool
+    $whatPageWeCameFrom = "historicalLookup";
+    $classN = $_POST["whichButtonHistoricalSearch"];
+    $classListDate = $_POST["input-date"];
+
+    $queryClass .= "and to_char(co.date, 'YYYY-MM-DD') = '{$classListDate}';";
+}
+else{ //shouldn't be here
+    echo "<h1>Please use 'recent classes' or the 'historical attendance lookup tool' to access attendance history.</h1>";
+    die; //on the Whirly Dirly
```
Comment (ghost) [inline views/attendance/historical_class_view.php]: `die()`

File: views/attendance/new_class.php [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        for(var k = 0; k < optionArray.length; k++){
+            selectTime.appendChild(optionArray[k]);
+        }
+    }
+
+    //input: hour and time integers
+    //output: option object
+    function createTime(hour, minute, amORpm){
+        var option = document.createElement('OPTION');
+        option.innerHTML = (hour + ":" + minute + " " + amORpm).toString();
```
Comment (ghost) [inline views/attendance/new_class.php:138]: `toString()` shouldn't be necessary here. `typeof(3+"AM")` is `"string"`

File: views/attendance/attendance_form.php (hunk 4 above)
Comment (Scott123180) [reply views/attendance/attendance_form.php:139]: firstClass is used to determine if a new record is going to be created for the person. When the form data is submitted, if firstclass is true, then a new person/participant is created in the system. If it is false, then that means we already found an id in the system for them.

Comment (ghost) [reply views/attendance/attendance_form.php:139]: Thanks for the clarification!

File: views/attendance/attendance_utilities.php (hunk 7 above)
Comment (Scott123180) [reply views/attendance/attendance_utilities.php]: yep, will use in V2 to validate the date

File: views/attendance/new_class.php (hunk 9 above)
Comment (Scott123180) [reply views/attendance/new_class.php:138]: for readability cause there's a lot of messed up typing in JS

**Comment by Scott123180** [comment]
implemented changes

**Review by blackjack26 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**